### ISSUE TITLE
Types error when creating a computed property based on another computed one

### ISSUE THREAD

**Issue by filipjnc** [ctrlplusb/easy-peasy#427]
TypeScript fails to compile in a scenario similar to the following one, where a computed property is based on another one: 

```
test('computed properties can access global state', () => {
  // arrange
  const store = createStore({
    products: {
      items: [{ id: 1, name: 'boots', price: 20 }],
      itemMap: computed([state => state.items], items =>
        items.reduce((acc, cur) => ({ ...acc, [cur.id]: cur }), {}),
      ),
      setProductName: action((state, payload) => {
        const product = state.items.find(p => p.id === payload.id);
        product.name = payload.name;
      }),
    },
    basket: {
      productIds: [1],
      products: computed(
        [
          (state, storeState) => storeState.products.itemMap,
          state => state.productIds,
        ],
        (productMap, productIds) => productIds.map(id => productMap[id]),
      ),
    },
  });
```

In my case the first computed property resolves to the object Role, only if the id is provided, otherwise it's undefined.

The second computed property uses the `Role `as dependency and returns a string containing `Role.name`.

```
interface Role {
  id: number,
  name: string,
  description: string
}
```

```
interface State {
  currentRole?: Computed<State, Role, StoreState>,
  anotherComputed?: Computed<State, string, StoreState>
}

export const model: State = {

  currentRole: computed(
    [
      (_, storeState) => storeState.role.currentId,
      (_, storeState) => storeState.data.rolesMap
    ], 
    (roleId, rolesMap) => !!roleId ? rolesMap[roleId] : undefined
  ),
  currentRoleFullName: computed(
    [
      state => state.currentRole,
    ],
    (role) => role && role.name
  ),
}
```
First it complains that the first computed cannot be assigned to the type because it can be undefined (I marked the property as optional with an `?`, so normally it should be no problem).
`Type 'Computed<State, Role | undefined, StoreState>' is not assignable to type 'Computed<State, Role, StoreState>'.`

If I change the type of the first computed to `Computed<State, Role | undefined, StoreState>` then it complains at the second, even if I check if role is not undefined:
`Property 'name' does not exist on type 'Computed<State, Role | undefined, StoreState>'.`

How do I handle this case? Why does the type resolve to Computed<> and not the underlying result type?

**Comment by ctrlplusb** [comment]
Hey @filipjnc 

Thanks for reporting this.

Would you mind creating a minimal TS example of this on CodeSandbox for me to debug.

Appreciate your time.

**Comment by filipjnc** [comment]
@ctrlplusb Somehow I cannot replicate the same error on CodeSandbox. I'll close it for now.

**Comment by ctrlplusb** [comment]
Awesome, thanks for this report @filipjnc

**Comment by ctrlplusb** [comment]
Hi @filipjnc 👋

So you actually uncovered two issues with the computed properties typings. 😅

I managed to resolve the one, which was that the result types for state resolver arrays were not being correctly inferred.

The other one has to do with making computed properties optional. This will unfortunately have to remain as "known issue". However, there is a very simple workaround.

I've included the fix for this in #459 and have updates the `computed` API docs to include notes on the known issue along with the workaround.

I'll include the snipped of that doc update below.

I'd appreciate it if you could try out the latest alpha and report back to me on your experience. 💜

---

### TypeScript: Defining a computed property as optional

Unfortunately, due to the way our typing system maps your model, you cannot declare a computed property as being optional via the `?` property postfix.

For example:

```typescript
interface StoreModel {
  products: Product[];
  totalPrice?: Computed<StoreModel, number>;
  //       👆
  // Note the optional definition
}

const storeModel: StoreModel = {
  products: [];
  // This will result in a TypeScript error 😢
  totalPrice: computed(
    state => state.products.length > 0
      ? calcPrice(state.products)
      : undefined
  )
}
```

Luckily there is a workaround; simply adjust the definition of your computed property to indicate that the result could be undefined.

```diff
  interface StoreModel {
    products: Product[];
-   totalPrice?: Computed<StoreModel, number>;
+   totalPrice: Computed<StoreModel, number | undefined>;
  }
```

**Comment by ctrlplusb** [comment]
Oh, and here is a fixed version of your CodeSandbox, which is utilising the latest alpha. 👍

https://codesandbox.io/s/falling-butterfly-9ftmt?file=/src/store/roleStoreModel.ts
